pkg/runtime/bulk_subscriber.go

```diff
@@ -0,0 +1,706 @@
+}
+
+// bulkSubscribeTopic subscribes to a topic for bulk messages and invokes subscriber app endpoint(s).
+
+// Major steps inside a bulk handler:
+//  1. Deserialize pubsub metadata and determine if rawPayload or not
+//     1.A. If any error occurs, send to DLQ if configured, else send back error for all messages
+//  2. Iterate through each message and validate entryID is NOT blank
+//     2.A. If it is a raw payload:
+//     2.A.i. Get route path, if processable
+//     2.A.ii. If route path is non-blank, generate base64 encoding of event data
+//     and set contentType, if provided, else set to "application/octet-stream"
+//     2.A.iii. Finally, form a child message to be sent to app and add it to the list of messages,
+//     to be sent to app (this list of messages is registered against correct path in an internal map)
+//     2.B. If it is NOT a raw payload (it is considered a cloud event):
+//     2.B.i. Unmarshal it into a map[string]interface{}
+//     2.B.ii. If any error while unmarshalling, send to DLQ if configured, else register error for this message
+//     2.B.iii. Check if message expired
+//     2.B.iv. Get route path, if processable
+//     2.B.v. If route path is non-blank, form a child message to be sent to app and add it to the list of messages,
+//  3. Iterate through map prepared for path vs list of messages to be sent on this path
+//     3.A. Prepare envelope for the list of messages to be sent to app on this path
+//     3.B. Send the envelope to app by invoking http endpoint
+//  4. Check if any error has occurred so far in processing for any of the message and invoke DLQ, if configured.
+//  5. Send back responses array to broker interface.
+func (a *DaprRuntime) bulkSubscribeTopic(ctx context.Context, policy resiliency.Runner,
+	psName string, topic string, route TopicRouteElem,
+) error {
+	ps, ok := a.pubSubs[psName]
+	if !ok {
+		return runtimePubsub.NotFoundError{PubsubName: psName}
+	}
+
+	req := pubsub.SubscribeRequest{
+		Topic:    topic,
+		Metadata: route.metadata,
+	}
+
+	bulkHandler := func(ctx context.Context, msg *pubsub.BulkMessage) ([]pubsub.BulkSubscribeResponseEntry, error) {
+		if msg.Metadata == nil {
+			msg.Metadata = make(map[string]string, 1)
+		}
+
+		msg.Metadata[pubsubName] = psName
+		bulkSubDiag := newBulkSubIngressDiagnostics()
+		bulkResponses := make([]pubsub.BulkSubscribeResponseEntry, len(msg.Entries))
+		rawPayload, err := contribMetadata.IsRawPayload(route.metadata)
+		if err != nil {
+			log.Errorf("error deserializing pubsub metadata: %s", err)
+			if dlqErr := a.sendBulkToDLQIfConfigured(ctx, psName, msg, route, nil, nil, &bulkSubDiag); dlqErr != nil {
+				populateAllBulkResponsesWithError(msg, &bulkResponses, err)
+				reportBulkSubDiagnostics(ctx, topic, &bulkSubDiag)
+				return bulkResponses, err
+			}
+			reportBulkSubDiagnostics(ctx, topic, &bulkSubDiag)
+			return nil, nil
+		}
+		routePathBulkMessageMap := make(map[string]pubsubBulkSubscribedMessage)
```

Comment:
Messages inside a bulk subscribe event can lead to different paths, due to match route algo etc., hence a map is required to maintain which path has got what messages to be delivered.